presto-main/src/main/java/io/prestosql/execution/scheduler/OptimizedLocalNodeSelector.java

```diff
@@ -0,0 +1,260 @@
+import com.google.common.collect.Multimap;
+import com.google.common.util.concurrent.ListenableFuture;
+import io.airlift.log.Logger;
+import io.prestosql.execution.NodeTaskMap;
+import io.prestosql.execution.RemoteTask;
+import io.prestosql.metadata.InternalNode;
+import io.prestosql.metadata.InternalNodeManager;
+import io.prestosql.metadata.Split;
+import io.prestosql.spi.HostAddress;
+import io.prestosql.spi.PrestoException;
+import org.apache.commons.math3.stat.descriptive.moment.StandardDeviation;
+
+import java.util.HashSet;
+import java.util.Iterator;
+import java.util.List;
+import java.util.Optional;
+import java.util.Set;
+
+import static io.prestosql.execution.scheduler.NodeScheduler.calculateLowWatermark;
+import static io.prestosql.execution.scheduler.NodeScheduler.randomizedNodes;
+import static io.prestosql.execution.scheduler.NodeScheduler.selectExactNodes;
+import static io.prestosql.execution.scheduler.NodeScheduler.selectNodes;
+import static io.prestosql.execution.scheduler.NodeScheduler.toWhenHasSplitQueueSpaceFuture;
+import static io.prestosql.spi.StandardErrorCode.NO_NODES_AVAILABLE;
+
+public class OptimizedLocalNodeSelector
+        extends SimpleNodeSelector
+{
+    private static final Logger log = Logger.get(OptimizedLocalNodeSelector.class);
+
+    OptimizedLocalNodeSelector(
+            InternalNodeManager nodeManager,
+            NodeTaskMap nodeTaskMap,
+            boolean includeCoordinator,
+            Supplier<NodeMap> nodeMap,
+            int minCandidates,
+            int maxSplitsPerNode,
+            int maxPendingSplitsPerTask)
+    {
+        super(nodeManager, nodeTaskMap, includeCoordinator, nodeMap, minCandidates, maxSplitsPerNode, maxPendingSplitsPerTask);
+    }
+
+    @Override
+    public SplitPlacementResult computeAssignments(Set<Split> splits, List<RemoteTask> existingTasks)
+    {
+        Multimap<InternalNode, Split> assignment = HashMultimap.create();
+        NodeMap nodeMap = this.nodeMap.get().get();
+        NodeAssignmentStats assignmentStats = new NodeAssignmentStats(nodeTaskMap, nodeMap, existingTasks);
+
+        ResettableRandomizedIterator<InternalNode> randomCandidates = randomizedNodes(nodeMap, includeCoordinator, ImmutableSet.of());
+        Set<InternalNode> blockedExactNodes = new HashSet<>();
+        boolean splitWaitingForAnyNode = false;
+        // splitsToBeRedistributed remains false for internal splits or if force-local-scheduling is true
+        boolean splitsToBeRedistributed = false;
+
+        //prioritize assignment of splits to local nodes when the local nodes have slots available in the first pass
+        Set<Split> splitsToBeScheduled = new HashSet<>(splits);
+        for (Split split : splits) {
+            if (split.isRemotelyAccessible() && (split.getAddresses().size() > 0)
+                    && split.getAddresses().stream().anyMatch(address -> !address.getHostText().equals("localhost"))) {
+                splitsToBeRedistributed = true;
```

Comment:
`splitsToBeRedistributed` should be set to true only if a node is selected right ?

Reply:
We are setting a this flag as true whenever splits enter the first stage (locality based assignment). We wanted the third stage (redistribution) to be invoked only when splits are entering the first stage, and that is what the `splitsToBeRedistributed` flag does. 

If splits are entering the 1st stage, they will remain unassigned in the 1st stage only when their local node is filled up. If only a few nodes are filled up after the 1st and the 2nd stage, leading to some splits remaining unassigned in the first stage (they would have been assigned in the second stage), redistribution will ensure a uniform workload. If all nodes are filled up, redistribution can not happen anyway.

Reply:
> they will remain unassigned in the 1st stage only when their local node is filled up

But they will remain unassigned even if their `candidatesNodes` are empty right ?

Reply:
Thanks for the suggestion. I have moved the location of where we are setting `splitsToBeRedistributed` to catch the case where splits have some addresses but `candidateNodes` are empty.